snuba/datasets/tags_column_processor.py

```diff
@@ -77,38 +125,135 @@ def __string_col(self, col: str) -> str:
         else:
             return "toString({})".format(escape_identifier(col))
 
-    def __tag_expr(self, column_name: str, table_alias: str = "",) -> str:
+    def __tag_expr(self, parsed_col: ParsedNestedColumn, table_alias: str = "",) -> str:
         """
         Return an expression for the value of a single named tag.
 
         For tags/contexts, we expand the expression depending on whether the tag is
         "promoted" to a top level column, or whether we have to look in the tags map.
         """
-        col, tag = NESTED_COL_EXPR_RE.match(column_name).group(1, 2)
         # For promoted tags, return the column name.
+        assert parsed_col.parameter
+        tag_name = parsed_col.parameter
+        col = parsed_col.col_name
         if col in self.__promoted_columns:
-            actual_tag = self.__get_tag_column_map()[col].get(tag, tag)
+            actual_tag = self.__get_tag_column_map()[col].get(tag_name, tag_name)
             if actual_tag in self.__promoted_columns[col]:
                 return qualified_column(self.__string_col(actual_tag), table_alias)
 
         # For the rest, return an expression that looks it up in the nested tags.
-        return u"{col}.value[indexOf({col}.key, {tag})]".format(
-            **{"col": qualified_column(col, table_alias), "tag": escape_literal(tag)}
+        return "{col}.value[indexOf({col}.key, {tag})]".format(
+            **{
+                "col": qualified_column(col, table_alias),
+                "tag": escape_literal(tag_name),
+            }
+        )
+
+    def __extract_top_level_tag_conditions(self, condition: Expression) -> List[str]:
```

Comment:
The behavior of this class is well illustrated by the test_tags_processor unit test.